Only allow one of each links attribute in resolver

Eh2406 opened the pull request:
This is a start on fixing the `links` part of #4902. It needs code that adds the links attribute to the index. But, I wanted to get feedback.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented:
Awesome, thanks!

It looks though like there's some CI failures? In addition to that, could some tests be added about successfully choosing a resolution when two possible graphs would have otherwise been possible?

The final comment I'd have for this is is that resolution tends to be a pretty performance sensitive path in Cargo, so would it be possible to avoid the `HashSet` construction on each iteration?

Eh2406 commented:
The CI failures are test looking for the old links checker, they will probably need to be updated. Maybe we should remove the old cheker entirely? That is a good suggestion for added test.

As to performance:
- Do we have a benchmark? 
- We could use a `Vec` as `n` is likely to be small.
- We could add it as a field of `Context`, but we would have to be careful it got updated in skink with `activations`.
- Why are we using the default hasher instead of fnv for all these hash maps?

Eh2406 commented:
We cant remove the old checker, ~~most~~all indexes don't have the new field. :-)

alexcrichton commented:
@Eh2406 yeah I thin kit's fine to remove the old checker entirely. Although I will say though that the error messages from the old checker are quite good, and it'd be a shame to lose those... Fixing that may require a more general overhaul to resolver error messages which already aren't the best unfortunately :(

In any case unfortunately I don't have many benchmarks for the resolver. The main one I've used in the past is "how long does it take to resolve Servo's dependency graph?" but without information in the index it'll be deceptively fast I think. In general though we just need to avoid allocations if possible on each turn of the loop if we can, or otherwise make sure the allocations are pretty small. Would it be possible to have a field that's updated over time rather than recreated each time?

The allocations have been most of the runtime so far, I've never seen the hashing part of a hash map show up in profiles.

Eh2406 commented:
The first time I tried this I had a persistent set, and got myself confused. Then I tried making the set dynamically and submit the pr. This time I got the persistent set to work. I think the third time's the charm!

As to the old checker, I don't think we can remove it until we are sure that all of the index (and clones of the index) are updated to have the new field. I could be wrong, I am new to this codebase. If I am wrong, please enlighten me!

Switching from the good old checker error messages to the resolver messages is a shame. But what can we do. :-(

How hard is it to resolve Servo's dependency graph in isolation as someone that has never touched Servo's code befor?

alexcrichton commented on src/cargo/core/resolver/mod.rs:
```diff
@@ -1047,6 +1053,9 @@ impl<'a> Context<'a> {
         if !prev.iter().any(|c| c == summary) {
             self.resolve_graph.push(GraphNode::Add(id.clone()));
+            if let Some(link) = summary.links() {
+                self.links.insert(link.to_owned());
```
Could this assert that we're not replacing somethign that was already there?

alexcrichton replied:
Hm that would indeed be bad! Thinking through how this can happen I think we could trigger this if an older cargo created a lock file with a bunch of new crates (that all have `links` in the manifest) and then a newer cargo tries to use that lock file.

In that case, sounds like we should skip the assert for now!

alexcrichton replied:
Er well, but if we do actually run into that case the previous project surely couldn't build previously becuase it would have been caught later... 

How about this, could we return an error from here to entirely abort resolution?

Eh2406 replied:
I think I have that working now.

alexcrichton commented:
Ok great! I think this looks good to me in terms of perf, and we can always continue to tweak as more cases come up.

I've actually found it somewhat easy to profile with Servo before, but unfortunately it may not be too interesting to profile with this PR as-is (as there's no `links` keys in the index). If you'd like though you could perhaps manufacture a `links` for all `Summary` instances created (a unique name in `links`) temporarily to test the performance?

Otherwise I've historically just cd'd into Servos' checkout and run `cargo generate-lockfile` or something like `cargo update -p servo` which shouldn't hit het network after the first time and stresseses the path of "we've already got a Cargo.lock so this shoudl be fast". Perhaps worth testing if you can with some synthetic `links` keys?

alexcrichton commented:
Oh right, some further thoughts:

* Let's actually not remove the old system just yet. Because many crates on crates.io don't have the `links` key in the index we won't otherwise catch those errors if we remove that logic. We can eventually remove it one day perhaps but only after we've published a lot.

* This will need to be coupled with an update to how we publish crates to the registry. That way information will actually go into the index!

Eh2406 commented:
So I got Servo set up so I can run the test. I benchmarked 4 versions. I am running windows so I don't have a `time` command. I am just looking at wall time for now. For each one I am running `cargo update -p servo`.

1. The base, the commit before my PR. Takes approximately 3 second to run. Is it usually this fast? Maybe I just got a lucky checkout of Servo.
2. The initial version of this PR, with a new `HashSet` for each loop. I killed it after a minute had already gone by. So, ya that is not acceptable.
3. The current head of this PR. This tests rolling out this PR without backfilling the index. Takes approximately 2 second to run.
4. If I add a line `links.or_else(|| if name.starts_with("sys-") {Some(name.to_string())} else {None})` to registry\index.rs. This tests after the index is filled in. Takes approximately 4 second to run.

alexcrichton commented:
Thanks for investigating @Eh2406! I'm actually surprised it took 3 seconds at all, that sort of update should happen in *milliseconds* ather than seconds... 

Well in any case sounds like the current PR as-is is certainly acceptable speed-wise!

Eh2406 commented:
I added the links attribute to the NewCrate struct then followed the error messages. Please let me know if I did that correctly.

alexcrichton commented:
Ok thanks @Eh2406! Thinking about this some more...

I'm getting more and more worried as I think about it about the error messages here. It's pretty well-known at this point I believe that the resolve error messages coming out of Cargo are pretty bad (almost un-actionable!). That's also coupled with the fact that non-resolvable crate graphs [look like hangs](https://github.com/rust-lang/cargo/issues/4066) I think could get us into a bit of a tight spot with this feature.

For example we've, since inception, had requests and [an implementation](https://github.com/rust-lang/cargo/pull/4515) for better error messages around the `links` key. That implies to me that this situation is actually coming up in the wild a nontrivial number of times. If we were to move the errors to graph resolution then I fear that all of these cases coming up in the wild will become un-debuggable. 

In that sense I'm not entirely sure how to balance this implementation. On one hand I definitely believe it's technically correct (although would be good to have a test to assert that this error can guide crate graph resolution). On the other hand though I'm not sure that this is worth the loss in error message quality. From my experience cases like https://github.com/rust-lang/cargo/issues/4902 are coming up pretty rarely where there's a resolvable crate graph but Cargo isn't finding it immediately.

I'm curious what you think about this as well?

Eh2406 commented:
I have a lot to say in response to your comment. How much will actually be relevant to this pr I won't know till I try and say it. But I hope to do a lot of work on the resolver and it will be relevant to that larger context. I will try and respond point by point.

> It's pretty well-known at this point I believe that the resolve error messages coming out of Cargo are pretty bad (almost un-actionable!).

I agree! I would love to work on fixing them when I understand the resolver deeply so I can see how to improve it. I started working on this PR so I could begin to understand the resolver. If I begin with some low hanging problems, even if I never understand the resolver deeply I will still have contributed.

I fear that making good error messages for resolver will be hard. Resolver currently returns some random crate it can not resolve, usually one that the user did not even know was being depend upon. It is not clear how to synthesize it into a larger coherent user centric message.

> that non-resolvable crate graphs look like hangs

I think this is surprisingly tightly tied to the error messages. I think the solution to failing fast on most non-resolvable crate graphs in valves taking the fact that some random crate can not be resolved synthesize it into a larger coherent problem then back track until the bigger problem goes away. If we had a good way of building a larger coherent problem we would have them both solved.

> For example we've, since inception, had requests and an implementation for better error messages around the links key. That implies to me that this situation is actually coming up in the wild a nontrivial number of times. If we were to move the errors to graph resolution then I fear that all of these cases coming up in the wild will become un-debuggable.

I think all the cases that are debugabal in the current system, would just resolve into something that worked. I.E. If you can fix it by specifying a specific version of a dependency, then it will just no longer be an issue. The process of backtracking will find the dependency that works.

Of course the resolver error messages are offil so if you have a bigger problem than pinning a dependency the error messages will get a lot worse.

> although would be good to have a test to assert that this error can guide crate graph resolution

I'd love some guidance on making a test in which cargo has multiple versions to choose from only the oldest of witch does not have a links attribute that conflicts with the main crate. Or any other test you would like to suggest.

> From my experience cases like #4902 are coming up pretty rarely where there's a resolvable crate graph but Cargo isn't finding it immediately.

I don't think that is an accurate description of #4902. I don't see that particular report mentioning cargo being slow. I see it reporting that cargo always builds a lock file that it then cannot use. With, in that case, no way to fix it. That is a really bad user experience.

I think over all this replaces friendly error messages that clearly tell you what problem cargo made for you with cargo silently not making the error in the first place.

Eh2406 commented:
I think this is a ergonomic win, but reasonable people can disagree. If you decide to postpone this until we somehow get better error messages then I will open up PRs related to other issues with the resolver and let this one bit rot. Knowing me, If I end up in a holding pattern I will likely get distracted, so one way or the other I'd like to keep things moving so I can contribute.

alexcrichton commented:
Thanks for the writeup @Eh2406! (and sorry for taking a few days to get back to you)

Mostly that all sounds great to me, I'd be thrilled to help you learn about the resolver if you'd like!

One point though is...

> I think over all this replaces friendly error messages that clearly tell you what problem cargo made for you with cargo silently not making the error in the first place.

Almost! A few days ago we talked about this more broadly in the cargo triage meeting and what to do with this PR. Our conclusion though was that we probably don't want to land this yet, but we're curious to learn more information!

My fear, more specifically, is that *entirely unresolvable* crate graphs will start showing up (producing bad error messages). I suspect that, because we've improved the error messages here, it's somewhat often coming up in practice that there's multiple native libraries of the same name that can't get built together. While in #4902 it's possible that Cargo could have actually found a successful resolution I suspect that is not the case for most of the times this comes up in practice. Whenever I've seen this before it's been impossible to resolve the crate graph with this constraint, but rather a dependency simply needed to be updated.

Along those lines though, @SergioBenitez  as the reporter of #4902 is this an issue that's coming up in practice for you a lot? The solution in this PR will solve #4902 but will unfortunately create some other problems as well, so we're trying to figure out how to best weigh the priorities here.

@Eh2406 if you'd like to tackle resolution one of the highest priority bits of it I think would be the error messages, and I'd be more than willing to help mentor you on improvements to that!

Eh2406 commented:
Just a note for when someone/I come back to this, in addition to the error messages, It will take some work to rebase/merge this with #4834, and that is a much bigger/more common UX improvement.

bors commented:
:umbrella: The latest upstream changes (presumably #4834) made this pull request unmergeable. Please resolve the merge conflicts.

Eh2406 commented:
I updated this to work with the new infrastructure from #5037, and added the hashmap from point 4. This now has fast backtracking and better error messages. Not exactly as good as the old ones, but close. 
This is on hold for the discussion in #5037 to finish.

Eh2406 commented:
Thanks for merging  #5037. I think this should now be pretty close to ready. Thoughts?

alexcrichton commented on tests/build-script.rs:
```diff
@@ -256,20 +257,18 @@ fn links_duplicates() {
+required by package `foo v0.5.0 ([..])`
+  multiple packages link to native library `a`, but a native library can be linked only once.
+  previously selected package `foo v0.5.0 ([..])`
+  possible versions to select: 0.5.0
```
I wonder if we could perhaps tidy up the wording around this sort of error? I found this sort of hard to follow as there's a lot of mentions of versions and such. Maybe something like:

```
error: failed to select a version for `a-sys`
  ... required by `foo v0.5.0`

the package `a-sys` links to the native library `a`, but it conflicts
with a previous package which links to `a` as well:

  foo v0.5.0
     .. required by ...

failed to select a version for `a-sys` which could resolve this conflict
```

I do realize that most of this is fitting in the mold of the previous error messages, but perhaps we can just tidy up the wording around them?

Eh2406 replied:
I will see what I can do. :-)

One thing to keep in mind is that different versions of a-sys may have different errors we need to report. For example only some versions have the link attribute. So we may, at least in theory, have to coherently report on a link error with 'a-fork-sys' and a semver error with 'a-sys' from before it had a link attribute.

Eh2406 commented:
Worked on the error messages, dose that look better?

bors commented:
:umbrella: The latest upstream changes (presumably #5063) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
@bors: r+

Looks amazing, thanks so much @Eh2406!

bors commented:
:pushpin: Commit 68a40ad has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 68a40ad42bd3aa0d45f39e587087a660a9e0c43c -->

bors commented:
:hourglass: Testing commit 68a40ad42bd3aa0d45f39e587087a660a9e0c43c with merge fe0c18b9dbd491845e5e3e50390dcfbe0cf00215...

alexcrichton commented:
@Eh2406 also, would you be up for following through with the crates.io changes needed here as well?

Eh2406 commented:
Um befor we merge, how do we test that adding `links` to the index will not break old versions of cargo?
@joshtriplett asked us [to test it explicitly](https://github.com/rust-lang/cargo/issues/4902#issuecomment-361361874).

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.3677), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/345742066?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing fe0c18b9dbd491845e5e3e50390dcfbe0cf00215 to master...

alexcrichton commented:
@Eh2406 oh I'd be fine adding a test yeah, although I'm quite confident that it won't require any code changes.

Eh2406 commented:
I don't think it's as simple as adding a test. Adding a test will test that it works with current cargo. I guess we would need to manually alter the index, and then every previous version of cargo.

Eh2406 commented:
Ok so I am trying to test if the links attribute will brake old cargo. This is a log of my progress.

1. set up a new project called testlinks, use `rustup override set 1.0.0-gnu` to make sure we are using an old cargo.
2. dell my `C:\Users\...\.cargo\registry` then run `cargo search winapi` inside testlinks. This checks out the index exactly as rust 1.0.0 wanted it.
3. Add a dependency to testlinks on ```cpython = {version = "*", features=["python27-sys"]}
python27-sys = "0.0.6"``` and generate a lock file. This set of dependencies are chosen to lead to having both ```[[package]]
name = "python27-sys"
version = "0.0.6"``` and ```[[package]]
name = "python27-sys"
version = "0.1.2"```
4. edit the index to add `,"links":"python27"}` to the end of each row of the `python27-sys` file. save and commit.
5. rerun `testlinks>cargo generate-lockfile` and insure it has the same result. 
6. try with a fresh build of cargo from master. witch should error but gives the same results, something is wrong. looks like something throw out my commit. re committing and still nothing.

This test failed! In 2 ways:
1. With cargo from 1.0.0 the `Updating registry` throws out my changes making the test useless.
2. With cargo from master it doesn't seem to be reading the attribute.

TLDR:
How do I get cargo from 1.0.0 to see an index with a links attribute? Why is cargo from master not reading the links attribute? Help!?

alexcrichton commented:
@Eh2406 IIRC Cargo at 1.0.0 actually unconditionally reads the registry, so you should be able to do:

1. Use Cargo 1.0.0 to generate a lock file
2. Edit the index. Take a crate that's in the lock file and add an `index` key to the row in the index for that crate
3. Rerun Cargo 1.0.0, it'll read the registry entry and ingore the `index` key

For master cargo are you using nightly or building from source? I think these changes haven't made their way into nightly yet unfortunately.

Eh2406 commented:
I think that is what I tried, but when Cargo 1.0.0 is `Updating registry` it does a clean checkout that removes my change to the index. I.E. change files to have the links and commit. Check that files are changed, And that commit in log. Run `testlinks>cargo generate-lockfile`. Index files do not have links, nor is my commit in the git log.

For the master test I am building from source.

alexcrichton commented:
Hm I'm not so sure in that case :(. Been awhile since I used Cargo 1.0.0... In any case using Cargo 1.20.0 or something like that (not too old) is probably sufficient. No one's using Rust 1.0.0 in production right now I'd imagine.

As to why master isn't reading the attribute... not sure!

Eh2406 commented:
Good news I figured it out, and all is well!
```cpython = {version = "*", features=["python27-sys"]} python27-sys = "0.0.6"``` Does have a solution it is to use an old version of `cpython` No error message needed.

Switching to ```cpython = {version = "0.1.0", features=["python27-sys"]}
python27-sys = "0.0.6"``` gives ```error: failed to select a version for `python27-sys`.``` as intended!

alexcrichton commented:
Nice!
